### ISSUE TITLE
Login fails if password has "<" or ":"

### ISSUE THREAD

**Issue by Atario** [erengy/taiga#170]
Don't know if something's broken on MAL's side or what, but for more than a day now, nothing I do will make Taiga connect to MAL — I always get "Authorization failed (invalid username or password)" messages, despite not changing them and the same credentials still working on their website.

**Comment by Atario** [comment]
Still not working.  For what it's worth, here's an HTTP session capture of the failure:

Request:

    GET http://myanimelist.net/api/account/verify_credentials.xml HTTP/1.1
    Host: myanimelist.net
    Proxy-Connection: Keep-Alive
    Accept: text/xml, text/*
    Accept-Charset: utf-8
    Accept-Encoding: gzip
    Authorization: Basic *[redacted]*
    User-Agent: api-taiga-32864c09ef538453b4d8110734ee355b

Response:

    HTTP/1.1 401 Unauthorized
    Cache-Control: no-cache
    Content-Type: text/html; charset=UTF-8
    Date: Sat, 29 Aug 2015 10:50:29 GMT
    Server: Apache
    Set-Cookie: MALHLOGSESSID=*[redacted]*; expires=Thu, 27-Aug-2020 10:50:29 GMT; Max-Age=157680000; path=/
    Set-Cookie: MALSESSIONID=*[redacted]*; expires=Sat, 05-Sep-2015 10:50:29 GMT; Max-Age=604800; path=/; HttpOnly
    Vary: Accept-Encoding
    WWW-Authenticate: Basic realm="MyAnimeList API"
    Content-Length: 19
    Connection: keep-alive
    Set-Cookie: visid_incap_81958=*[redacted]*; expires=Mon, 28 Aug 2017 09:40:31 GMT; path=/; Domain=.myanimelist.net
    Set-Cookie: nlbi_81958=*[redacted]*; path=/; Domain=.myanimelist.net
    Set-Cookie: incap_ses_257_81958=*[redacted]*; path=/; Domain=.myanimelist.net
    X-Iinfo: *[redacted]* q(0 0 0 -1) r(1 1) U13
    X-CDN: Incapsula
    Proxy-Connection: keep-alive

    Invalid credentials

**Comment by erengy** [comment]
Could you try decoding the `Authorization` header (it's base64 encoded) and see if it translates to your `username:password`?

**Comment by Atario** [comment]
Yeah, it does.

However, doing this prompted me to try something.  See, my password is machine-generated, and is long and contains unusual characters (nothing outside printable ASCII though).  So I:

* Changed it to a weaker one (shorter, having only alphanumerics).  This made Taiga able to connect again.  Hmm.
* Put it back to the original password, and Taiga *still* worked.  Odd.
* Went to my other machine which I hadn't touched yet and therefore still had the original password.  That still *didn't* work.  Strange.
* Went through the same steps and it did the same thing as the first machine.  So at this point the password was the hard one.
* Exited and reopened the first machine's Taiga instance.  It stopped working again.
* Exited and reopened the second machine's Taiga instance.  It stopped working again too.
* Got the first machine working again with the hard password, then changed the MAL password to the easy one without changing it in Taiga (so now it really has the wrong password).  Taiga still worked.
* Changed MAL password back to the hard one.  Taiga still worked.

So, it seems like only the first time Taiga tries to communicate after starting really counts, and after that it hangs on to some auth token or something till you exit, at which point it's discarded again.

But that first time, Taiga *can't* authenticate if certain characters are in the password.

At this point, I exhaustively ran through the hard password, eliminating non-alphanum characters one by one till Taiga worked.  Turns out the offending character was "<".  On the other hand, that was the last non-alphanum character and by that point I was down to a fully alphanum password.  So who knows if the other non-alphanum characters are ok or not…

**tl;dr:**  Put "<" anywhere in your password and Taiga can't log in to MAL, though the website itself is fine with it.  Unknown if other characters also cause this.

**Comment by erengy** [comment]
Thanks for the detailed explanation! I assume that this issue is related to MAL's recent security measures, which resulted in yet another bug for MAL API clients such as Taiga. I'm currently busy with work-related stuff, but I'll investigate this issue when I have the time and report it to MAL if necessary.

**Comment by erengy** [comment]
Reported the issue:
http://myanimelist.net/forum/?topicid=1423799

**Comment by erengy** [comment]
Just checked this again and the issue remains unfixed.

**Comment by rr-** [comment]
Oh boy, I'm so glad I'm through with their API...

**Comment by erengy** [comment]
The same issue arises if you have a `:` in your password, even though [RFC2617](https://www.ietf.org/rfc/rfc2617.txt) allows the character. Probably because they're splitting `username:pass:word` as `username` `pass` `word` rather than `username` `pass:word`.
